### PR TITLE
More reliable shutdown of the EnsoContext to save resources

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

There is 572 threads during `RuntimeServerTest` execution create. They stay around because the context isn't closed after each test finishes. Adding `afterEach` cleanup sequence to `runtime-with-instruments` tests. Improving robustness of overall `EnsoContext` [shutdown](https://github.com/enso-org/enso/pull/6468#discussion_r1180348303).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- All code has been tested:
  - [x] Unit tests are still working

### CODE REVIEW

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 1]
```diff
@@ -331,6 +331,10 @@ class RuntimeServerTest
     val Some(Api.Response(_, Api.InitializedNotification())) = context.receive
   }
 
+  override protected def afterEach(): Unit = {
+    context.executionContext.context.close()
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:335]: This helps the "engine threads", but the test should also stop Akka, @4e6. Dmitry, if you know how to do that feel free to directly commit to this branch.

Comment (4e6) [reply engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:335]: We're not using Akka in `runtime*` projects. Only in the `language-server` and `project-manager`

**Comment by 4e6** [comment]
@JaroslavTulach do you have any estimate of how much resources can occupy the test Context? I'm just curious.

**Comment by JaroslavTulach** [comment]
> @JaroslavTulach do you have any estimate of how much resources can occupy the test Context? I'm just curious.

I am using VisualVM to _Monitor_ the started `test *RuntimeServerTest` JVM. There used to be 572 threads at the end of `RuntimeServerTest` execution:
Original
now there is at most 39 threads:
After the fixes
each thread occupies at least 1MB because it pre-allocates a stack. That's at least 572MB of system resources. In adition to that the JVM is using 700MB of heap (that hasn't changed). E.g. we can save 1/3 of the RSS size of the process when running the new version.

**Comment by 4e6** [comment]
Not bad :thinking:

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala (hunk 1 above)
Comment (4e6) [reply engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:335]: I was wrong, in fact, runtime tests are using Akka.

File: engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeExecutionEnvironmentTest.scala [hunk 2]
```diff
@@ -129,6 +129,11 @@ class RuntimeExecutionEnvironmentTest
     val Some(Api.Response(_, Api.InitializedNotification())) = context.receive
   }
 
+  override protected def afterEach(): Unit = {
+    context.executionContext.context.close()
```
Comment (JaroslavTulach) [inline engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeExecutionEnvironmentTest.scala:133]: When I added all these `afterEach` cleanups, running the whole `sbt runtime-with-instruments/test` started to randomly fail. Turned out I had to also f71557db - e.g. wait for the `shutdown` of the `EnsoContext` threads.

**Review by Akirathan — APPROVED** [review]

**Review by hubertp — APPROVED** [review]

**Pull request merged**